packages/fxa-auth-server/lib/pushbox/db.ts

```diff
@@ -24,6 +24,42 @@ export class PushboxDB {
     });
     return res;
   }
+
+  async retrieve(x: {
+    uid: string;
+    deviceId: string;
+    limit: number;
+    index?: number | null;
+  }) {
+    const lastest = await Record.query()
+      .select('idx')
+      .findOne({ user_id: x.uid, device_id: x.deviceId })
+      .orderBy('idx', 'desc');
+    const maxIndex = lastest?.idx || 0;
+    const query = Record.query()
+      .where({
+        user_id: x.uid,
+        device_id: x.deviceId,
+      })
+      .where('ttl', '>=', Math.floor(Date.now() / 1000))
+      .limit(x.limit)
+      .orderBy('idx');
+    if (x.index) {
+      query.where('idx', '>=', x.index);
+    }
+
+    const messages = await query.execute();
+    // because there was an ORDER BY idx in the query
+    const lastIndex =
+      (messages.length && messages[messages.length - 1].idx) || 0;
```

Comment:
aside: If we're using Node 16.6.0+, I think we can also use the new [`Array#at()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at) API, and do `users.at(-1)?.idx || 0`

Reply:
Thanks!  If I'm going to make other changes for this PR I'll include it.  Otherwise I'll include it in anther PR for this epic.

Reply:
Maybe not worth it.
Per package.json `engines.node`, we might have a minimum supported version of 16.12.x. But I guess it depends on what the Node version number the SRE team is deploying to stage/prod on. 🤷 

https://github.com/mozilla/fxa/blob/2700d76c06e93aba05836bd3b77e2f543769fd6e/package.json#L43-L45